### PR TITLE
Provide an interceptor for Data Provider

### LINKED ISSUE #2111: Provide an interceptor for Data Provider
`7.0.0-beta7`

### Expected behavior
TestNG should provide an interceptor that can intercept data provider returned values for alteration.

### Actual behavior

Nothing exists which supports this yet.

### Is the issue reproductible on runner?

- [X] Shell
- [X] Maven
- [X] Gradle
- [ ] Ant
- [ ] Eclipse
- [X] IntelliJ
- [ ] NetBeans

More details are available in this stack overflow question : https://stackoverflow.com/q/56972050/679824

### PR DESCRIPTION
**Pull request by krmahadevan**
Closes #2111

Fixes #2111 .

### Did you remember to?

- [X] Add test case(s)
- [X] Update `CHANGES.txt`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

### CODE REVIEW

File: src/main/java/org/testng/DataProviderHolder.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import java.util.Collections;
+import org.testng.collections.Sets;
+
+/**
+ * A holder class that is aimed at acting as a container for holding various different aspects of a
+ * data provider such as listeners/interceptors etc.,
+ */
+public class DataProviderHolder {
+
+  private Collection<IDataProviderListener> listeners = Sets.newHashSet();
```
Comment (juherr) [inline src/main/java/org/testng/DataProviderHolder.java]: could be final

**Review by juherr — COMMENTED** [review]
Awesome job, you are in fire theses days 👍

File: src/main/java/org/testng/DataProviderHolder.java (hunk 1 above)
Comment (krmahadevan) [reply src/main/java/org/testng/DataProviderHolder.java]: Sure. I will fix this. @juherr - Any other observations or can I consider this approved and can proceed with getting this merged ?

Comment (juherr) [reply src/main/java/org/testng/DataProviderHolder.java]: no, everything else looks good 👍

**Pull request merged**